### ISSUE TITLE
Tags for Headings

### ISSUE THREAD

**Issue by wfreeman-aph** [daisy/ebraille#66]
This should be built in. You have \<h1> through \<h6> in HTML for headings.

BANA regions have four different headings that would likely require these tags. Running Headings which can appear on nearly every page and have the title of the book, then centered headings, cell-5 headings, and cell-7 headings. Not every book has a running heading, so it might be odd to make that the h1. Centered headings seem like the best candidate for h1s since they signify the main heading of a section. Cell-5 and Cell-7 are sub-headings, so would be good candidates for h2 and h3 respectively.

Questions: Do we need special elements for running headings and maybe use a \<span> instead of one of the heading tags, or could you make them an h4? What do other braille regions require here? Do any braille regions require more than four headings? Do we need special elements like classes or attributes for any headings?

**Comment by LeonaHolloway** [comment]
From Australia: We definitely think of the running heading as something very different from a level heading. A special element for running headings would be preferred. We have four levels of headings.

**Comment by mattgarrish** [comment]
The DPUB-ARIA 1.1 revision includes new [`doc-pageheader`](https://www.w3.org/TR/dpub-aria-1.1/#doc-pageheader) and [`doc-pagefooter`](https://www.w3.org/TR/dpub-aria-1.1/#doc-pagefooter) roles. These are meant to hide the repetitive content from AT while reading (e.g., in google docs), but probably could be used to provide a running head for embossing.

But the revision is only at the working draft stage, so there's no guarantee the roles will make it to the final recommendation. This would be an interesting case to prove the roles will be used, though, which is the big hurdle they still face.

**Comment by mixa72** [comment]
From Switzerland: We also use h1 to h6 for the section headings and a running footer between the print page number (left) and the braille page number (right) to display the current section heading, see an example [here](http://snaekobbi.github.io/requirements/v1.5/test/page-layout/test.xprocspec.xhtml#test6)

**Comment by GouldenJen** [comment]
Do we need to include parameters here for each of these, such as h2: blocked in cell 5, preceded by a blank line? Or are we just at the point of determining which tags will apply to which braille format elements?

**Comment by GouldenJen** [comment]
I realize these will vary by country.

**Comment by skntkacm** [comment]
No need for more than 4 headings in Slovakia. 
The problem which occurs now is blank line before and after a heading. As I mentioned already on one of our meetings:

- if a heading tag will include blank line before and after it, it is crucial to see it immediatelly in the e-braille file as focusable blank lines.
- if a general heading tag would not necessarily include blank lines, there could be specific headings, such as "BANA H1", "BANA H2" ... - some specific style closely related to the specific rule strictly demanding a blank line.

**Comment by jrbowden** [comment]
In the UK we have three main kinds of heading, but each has a variant, so yes, let's have the standard 6 levels. The CSS can determine the precise formatting of each level for each job.
We use running page headers in the UK. Optional print page number on left, running head centred and braille page number on the right.
The issue of blank lines needed also occurs in the UK, again this could be handled by CSS class.

**Comment by jrbowden** [comment]
HTML5 also has article and section tags which might be helpful?

**Comment by wfreeman-aph** [comment]
@GouldenJen I think these tickets can just be about which tags are used. CSS will determine how each tag is formatted and those will, as you rightly point out in your other comment, vary by region.

**Comment by wfreeman-aph** [comment]
Discussed in meeting, by James: "header is best fit for running headings." Michael Hunsaker mentioned that there are different ways of approaching it and you can just define "running head" as its own thing and define it that way.

There's also the potential for different running heads for different line lengths. James recommends header tag and CSS to specify the line lengths.

George recommends that the heading hierarchy should follow the hierarchy in the book for improved navigation and accessibility. William suggests that this might be too big for the group to decide. Some discussion follows about how braille is created vs making an ebook from scratch. This includes the idea of not fixing the mistakes in the original print. We may not be able to validate this aspect of the file type as the transcriber may have to break the expected hierarchy to accurately represent the book.

Question of whether we need more than 6 headings. George says flatten it after 6 and just use attributes to represent that.

Charles mentioned running headers. Embossing vs refreshable and the experience being different and he wondered if that was reflected in a ticket. William was thinking it was but would check. Thinking it was in the hide/show ticket.

Consensus is using header for running heads and h1-h6 would be used for all headings within the book.

**Comment by mattgarrish** [comment]
The creation of running headers here still sounds problematic. How would CSS be able to template the running headers if they're contingent on a print page break in the text, for example? I'd imagine only the reading system could do this.

Even using CSS counters for braille page numbers sounds complicated when dealing with multi-document files. Does each chapter's braille page numbering begin again at 1 each time, or how would numbering stay consistent across documents without reading system intervention?

**Comment by jrbowden** [comment]
Slightly speculating, I imagine the actual running header text should usually be specified alongside or after a heading, not at each print page change.
See Wiki example.

I think, quite how print and braille page numbers get added to form a "Page Information Line" is still for discussion.

A page information line typically contains (in the UK) print page number, running head and braille page number).

I imagine the reading system will have to be capable of generating (braille) numbers and know various things, including:
* Numbering scheme for braille pages
(we'll need to define these)
* Numbering scheme for continuation pages (what happens when one print page spans multiple braille pages)
(we'll need to specify these)
* Numbering scheme when multiple print pages span one braille page
(we'll need to specify these)

All of this is,  I imagine, going to be extension to standard CSS.

e.g. for the basic braille page numbers, we possibly need:
* Numbering scheme: Arabic, Roman, (alphabetic) etc
* Braille numbering scheme (Braille, Antoine etc)
* With or without number/letter prefix (probably)
* Any prefix needs (e.g. BANA p and t pages)
* Position (left, right, centre, outer, inner, off; every page, odd only, even only, etc)
* Starting number (over the entire book, per volume, change at a particular point etc)

-- I reckon a lot still to discuss.

**Comment by mattgarrish** [comment]
> All of this is, I imagine, going to be extension to standard CSS.

I'm concerned this is going to prove to be more complicated than CSS can handle. The dynamic aspects of what's needed would seem to me to go beyond what an `@page` rule could accomplish, if that's the idea for handling these.

You can't really use the `content` property to get the needed text to inject into the page header, for example, as having ever-changing headers means the one closest to (or on) the paginated page has to be used, but what CSS selector would find that to add it? The same applies to getting the print page breaks in the text.

There's also the problem that if you only want to print one document, you presumably have to prep all the documents prior to it to find out the correct braille starting page. And even if you print the whole document the reading system has to override the CSS to inject the correct starting counter.

This feature may prove highly dependent on the reading system finding and injecting the correct information, in which case the `header` element might have to define a template for the reading system to pattern the text on.

**Comment by bertfrees** [comment]
> How would CSS be able to template the running headers if they're contingent on a print page break in the text

@mattgarrish Can you explain what exactly you mean by this?

> Even using CSS counters for braille page numbers sounds complicated when dealing with multi-document files. Does each chapter's braille page numbering begin again at 1 each time, or how would numbering stay consistent across documents without reading system intervention?

Good point. I think the reading system indeed needs to intervene to ensure consistency of the braille page counter(s) across documents.

> I imagine the reading system will have to be capable of generating (braille) numbers and know various things, including: [...]

The reading system will have to manage (manipulate, maybe restrict) braille page counters a bit, but formatting could be left entirely up to CSS. Everything you mention regarding braille page numbers can be done with standard CSS. To handle print page numbers with CSS, some extensions are indeed needed.

> Numbering scheme for braille pages

Formatting braille page numbers should be possible through (custom) [counter styles](https://www.w3.org/TR/css-counter-styles-3/).

> Numbering scheme for continuation pages

In https://github.com/daisy/ebraille/issues/65 I included an example style sheet for continuation pages. It uses the `string-set` property and `string()` function though which are not standard CSS (see below).

> Numbering scheme when multiple print pages span one braille page

DAISY Pipeline handles this, as well as running headers, entirely with the `string-set` property and the `string()` function, from the [CSS Generated Content for Paged Media](https://www.w3.org/TR/css-gcpm-3/#named-strings) working draft. [Named strings](https://www.w3.org/TR/css-gcpm-3/#named-strings) is a feature that was invented exactly for these use cases. Unfortunately the draft never made it into the standard.

Note that in [the braille CSS spec](http://braillespecs.github.io/braille-css/#h3_named-strings) I made the `string()` function more powerful to handle all print page number requirements I have come across.

**Comment by mattgarrish** [comment]
> > How would CSS be able to template the running headers if they're contingent on a print page break in the text
> 
> @mattgarrish Can you explain what exactly you mean by this?

The static print page numbers from the source are marked up in the document so you have to retrieve the one(s) that most closely correspond to the paginated braille (I assume you may sometimes have two numbers if the braille page crosses two print pages). You can't know what page break to select until the braille pagination has happened, so it seems like you need page-based CSS selectors to get the right information. Nothing like that exists, does it?

(This strikes me as the more complex problem of the two.)

> In #65 I included an example style sheet for continuation pages. It uses the `string-set` property and `string()` function though which are not standard CSS (see below).

Right, this is what I was getting at in terms of the pagination having to be passed to each document. It sounds more workable if the entire publication is printed. But if you're only printing one document out of the whole, you seemingly have to load and paginate every document that precedes the one you want in order to get the starting sequence number. That seems like an expensive operation for one document.

But maybe this particular problem is out of scope. What constitutes a single document is not going to be readily understandable to the average user. Assuming the document breaks are seamless to users, like with an EPUB, printing the "current document" could lead to all kinds of outcomes depending on how it is chunked under the hood. If reading systems offer the ability to print, perhaps the expectation should be to print the entirety.

**Comment by bertfrees** [comment]
Ok, I see what you mean now. Indeed, in some cases you can't know what print page break / header to select until the braille pagination has happened, at least partially, and nothing like page-based CSS selectors exists. But there is the above-mentioned "named strings" feature, or some other CSS extension. There is of course also the option to leave it up to the reading system.

> Right, this is what I was getting at in terms of the pagination having to be passed to each document.

I don't quite see the connection with continuation pages and how DAISY Pipeline handles those.

But anyhow, having to paginate a part or the entirety of the book is of course inevitable if you want to render the current braille page, or the braille page of a referenced item in the table of contents for example. It is expensive, but I don't see an alternative. If the book is not paginated, it does not make sense to render braille page numbers.

The same problem exists (though to a much lesser extend) for print page numbers when the print page number format depends on the pagination, such as when you want to show the range of print pages on a braille page or spread, or if you want to keep count of the number of braille pages the print page spans. Again, these formats don't make sense when the book is not paginated.

> If reading systems offer the ability to print, perhaps the expectation should be to print the entirety.

This is exactly what I'm thinking.

**Comment by mattgarrish** [comment]
> But there is the above-mentioned "named strings" feature, or some other CSS extension.

We could use an @page rule to template where the print page numbers go (e.g., define some value to replace with the numbers), but I'm thinking it's going to require some logic for the reading system on how to calculate the pagination to place.

Only giving it quick thought, but there are at least a few scenarios for adding the source pagination to the header:

- if the paginated braille page starts with a print page marker, use that marker (e.g., a new chapter, assuming printing forces braille page breaks)
  - if there's another print page marker on the page, there needs to be a range
- if the paginated braille page does not start with a print page marker, scan back to find the last one encountered
  - if there's another print page marker later on the page, there needs to be a range
- if there's no print page marker on the page, scan back to find the last one encountered

I don't know how hard it will be to implement this kind of logic within a css processor, however.

**Comment by bertfrees** [comment]
> We could use an `@page` rule to template where the print page numbers go (e.g., define some value to replace with the numbers), but I'm thinking it's going to require some logic for the reading system on how to calculate the pagination to place.

DAISY Pipeline shows that the desired behavior of this logic could be defined entirely in CSS, through the [`string()` function](http://braillespecs.github.io/braille-css/#h4_the-string-function). This function can handle all the scenarios you mentioned and all other scenarios I have come across.

**Comment by mattgarrish** [comment]
Okay, I'll trust your experience on this!

**Comment by bertfrees** [comment]
It might indeed prove to be tricky to implement within a CSS processor though.

**Comment by mattgarrish** [comment]
It's not clear to me if a reading system is also a "printing system", or if it's envisioned that these are separate processes. It could be that the details of how to paginate and add the running headers are out of scope for this specification if the idea is only that an ebraille file can also be an input to transcription software (in which case, the author might not need to concern themselves with the format of running headers, as the transcription software probably handles that part). In other words, maybe a CSS processor isn't involved in the printing process.
